docs/streaming/README.md

```diff
@@ -0,0 +1,100 @@
+
+# Event Streaming
+
+Event streaming is a new asynchronous RPC mechanism that is being added to Consul. Instead
+of synchronous blocking RPC calls (long polling) to fetch data when it changes, streaming
+sends events as they occur, and the client maintains a materialized view of the events.
+
+At the time of writing only the service health endpoint uses streaming, but more endpoints
+will be added in the future.
+
+## Overview
+
+The diagram below shows the components that are used in streaming, and how they fit into
+the rest of Consul.
+
+![Streaming Overview](./overview.svg)
+
+<sup>[source](./overview.mmd)</sup>
+
+Read requests are received either from the HTTP API or from a DNS request. They use
+[rpcclient/health.Health]
+to query the cache. The [StreamingHealthServices cache-type] uses a [materialized view]
+to manage subscriptions and store the aggregated events. On the server, the
+[SubscribeEndpoint] subscribes and receives events from [EventPublisher].
+
+Writes will likely enter the system through the client as well, but to make the diagram
+less complicated the write flow starts when it is received by the RPC endpoint. The
+endpoint calls raft.Apply, which if successful will save the new data in the state.Store.
+When the [state.Store commits] it produces an event which is managed by the [EventPublisher]
+and sent to any active subscriptions.
+
+[rpcclient/health.Health]: https://github.com/hashicorp/consul/blob/master/agent/rpcclient/health/health.go
+[StreamingHealthServices cache-type]: https://github.com/hashicorp/consul/blob/master/agent/cache-types/streaming_health_services.go
+[materialized view]: https://github.com/hashicorp/consul/blob/master/agent/submatview/materializer.go
+[SubscribeEndpoint]: https://github.com/hashicorp/consul/blob/master/agent/rpc/subscribe/subscribe.go
+[EventPublisher]: https://github.com/hashicorp/consul/blob/master/agent/consul/stream/event_publisher.go
+[state.Store commits]: https://github.com/hashicorp/consul/blob/master/agent/consul/state/memdb.go
+
+
+## Event Publisher
+
+The [EventPublisher] in at the core of streaming. It receives published events, and
+subscription requests, and forwards events to the appropriate subscriptions. The diagram
+below illustrates how events are stored by the [EventPublisher].
+
+![Event Publisher layout](./event-publisher-layout.svg)
+
+<sup>[source](./event-publisher-layout.mmd)</sup>
+
+When a new subscription is created it will create a snapshot of the events required to
+reflect the current state. This snapshot is cached by the [EventPublisher] so that other
```

Comment:
Maybe worth setting expectation that this cache is very short lived here and mostly for de-duplication of concurrent new streams (i.e. thundering herds) than a general purpose cache?

Reply:
Ya, that might be good to include. In these high level overview docs I generally try to keep the things pretty light, so that the docs are less likely to get out of sync with the code. I'm not sure how likely it is that we would change the TTL. I'll add some more detail here.